File: screenshot/View/Screenshot/ScreenshotsNavigationController.m
//
//  ScreenshotsNavigationController.m
//  screenshot
//
//  Created by Corey Werner on 8/8/17.
//  Copyright © 2017 crazeapp. All rights reserved.
//

#import "ScreenshotsNavigationController.h"
#import "ProductsViewController.h"
#import "screenshot-Swift.h"

#import <FBSDKCoreKit/FBSDKCoreKit.h>

@interface ScreenshotsNavigationController () <UIStateRestoring, ViewControllerLifeCycle, ScreenshotsViewControllerDelegate, NetworkingIndicatorProtocol>

@property (nonatomic, strong) ScreenshotPickerNavigationController *pickerNavigationController;
@property (nonatomic, strong) ClipView *clipView;

@property (nonatomic, strong, nullable) Class previousDidAppearViewControllerClass;

@property (nonatomic, strong) NSNumber *restoredScreenshotNumber;

@end

@implementation ScreenshotsNavigationController

#pragma mark - Life Cycle

- (instancetype)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil {
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        self.restorationIdentifier = NSStringFromClass([self class]);
        
        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(coreDataStackCompleted:) name:[NotificationCenterKeys coreDataStackCompleted] object:nil];
        
        _screenshotsViewController = ({
            ScreenshotsViewController *viewController = [[ScreenshotsViewController alloc] init];
            viewController.delegate = self;
            viewController.lifeCycleDelegate = self;
            viewController.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed:@"NavigationBarAddPhotos"] style:UIBarButtonItemStylePlain target:self action:@selector(presentPickerViewController)];
            viewController.navigationItem.rightBarButtonItem.tintColor = [UIColor crazeRed];
            viewController;
        });
        
        self.viewControllers = @[self.screenshotsViewController];
        
        [AssetSyncModel sharedInstance].networkingIndicatorDelegate = self;
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.view.backgroundColor = [UIColor background];
}

- (void)coreDataStackCompleted:(NSNotification *)notification {
    if (self.restoredScreenshotNumber) {
        ProductsViewController *productsViewController = (ProductsViewController *)self.topViewController;
        
        if ([productsViewController isKindOfClass:[ProductsViewController class]]) {
            productsViewController.screenshot = [self.screenshotsViewController screenshotAtIndex:self.restoredScreenshotNumber.integerValue];
            self.restoredScreenshotNumber = nil;
        }
    }
}

- (void)dealloc {
    [AssetSyncModel sharedInstance].networkingIndicatorDelegate = nil;
    
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}


#pragma mark - View Controller Life Cycle

- (void)viewController:(UIViewController *)viewController didAppear:(BOOL)animated {
    if (viewController == self.screenshotsViewController &&
        self.previousDidAppearViewControllerClass == [ProductsViewController class])
    {
        if ([self needsToPresentPickerViewController]) {
            // Allow the view controller transition view to cleanup
            dispatch_async(dispatch_get_main_queue(), ^{
                [self presentPickerClipView];
            });

            // Go back into Products before presenting the next view
            self.previousDidAppearViewControllerClass = nil;

        } else if ([self needsToPresentPushAlert]) {
            [self presentPushAlert];
        }
    }
    
    self.previousDidAppearViewControllerClass = [viewController class];
}


#pragma mark - Screenshots

- (void)screenshotsViewController:(ScreenshotsViewController *)viewController didSelectItemAtIndexPath:(NSIndexPath *)indexPath {
    Screenshot *screenshot = [viewController screenshotAtIndex:indexPath.item];
    
    ProductsViewController *productsViewController = [self createProductsViewController];
    productsViewController.screenshot = screenshot;
    [self pushViewController:productsViewController animated:YES];
    
    if (screenshot.isNew) {
        [screenshot setViewed];
    }
    
    [[RatingFlow sharedInstance] recordSignificantEvent];
}

- (void)screenshotsViewControllerDeletedLastScreenshot:(ScreenshotsViewController *)viewController {
    if ([self needsToPresentPickerViewController]) {
        [self presentPickerClipView];
    }
}

- (void)screenshotsViewControllerWantsToPresentPicker:(ScreenshotsViewController *)viewController {
    [self presentPickerViewController];
}


#pragma mark - Products

- (ProductsViewController *)createProductsViewController {
    ProductsViewController *viewController = [[ProductsViewController alloc] init];
    viewController.lifeCycleDelegate = self;
    viewController.hidesBottomBarWhenPushed = YES;
    return viewController;
}


#pragma mark - Screenshot Picker

- (ScreenshotPickerNavigationController *)createScreenshotPickerNavigationController {
    ScreenshotPickerNavigationController *navigationController = [[ScreenshotPickerNavigationController alloc] init];
    navigationController.cancelButton.target = self;
    navigationController.cancelButton.action = @selector(pickerViewControllerDidCancel);
    navigationController.doneButton.target = self;
    navigationController.doneButton.action = @selector(pickerViewControllerDidFinish);
    return navigationController;
}

- (BOOL)needsToPresentPickerViewController {
    return ![[NSUserDefaults standardUserDefaults] boolForKey:UserDefaultsKeys.onboardingPresentedScreenshotPicker];
}

- (void)presentPickerViewController {
    [self dismissPickerClipView];
    
    ScreenshotPickerNavigationController *picker = [self createScreenshotPickerNavigationController];
    self.pickerNavigationController = picker; // ???: is this needed?
    
    [self presentViewController:picker animated:YES completion:nil];
    
    [[NSUserDefaults standardUserDefaults] setBool:YES forKey:UserDefaultsKeys.onboardingPresentedScreenshotPicker];
    [[NSUserDefaults standardUserDefaults] synchronize];
    
    [AnalyticsTrackers.standard track:@"Opened Picker" properties:nil];
}

- (void)pickerViewControllerDidCancel {
    [self dismissViewControllerAnimated:YES completion:^{
        if ([self needsToPresentPushAlert]) {
            [self presentPushAlert];
        }
    }];
}

- (void)pickerViewControllerDidFinish {
    // Remove picker to reset state
    self.pickerNavigationController = nil;
    
    [self dismissViewControllerAnimated:YES completion:nil];
}


#pragma mark - Push Permission

- (BOOL)needsToPresentPushAlert {
    return ![[NSUserDefaults standardUserDefaults] boolForKey:UserDefaultsKeys.onboardingPresentedPushAlert] && [[PermissionsManager shared] _hasPhotoPermission];
}

- (void)presentPushAlert {
    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"Start Screenshotting" message:@"Open your favorite apps and take screenshots of photos with clothes, then come back here to shop them!" preferredStyle:UIAlertControllerStyleAlert];
    [alertController addAction:[UIAlertAction actionWithTitle:@"Ok" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        [[PermissionsManager shared] _requestPushPermissionWithResponse:^(BOOL granted) {
            if (granted) {
                [self.screenshotsNavigationControllerDelegate screenshotsNavigationControllerDidGrantPushPermissions:self];
                
                [AnalyticsTrackers.standard track:@"Accepted Push Permissions" properties:nil];
                
            } else {
                [AnalyticsTrackers.standard track:@"Denied Push Permissions" properties:nil];
            }
        }];
    }]];
    [self presentViewController:alertController animated:YES completion:nil];
    
    [[NSUserDefaults standardUserDefaults] setBool:YES forKey:UserDefaultsKeys.onboardingPresentedPushAlert];
    [[NSUserDefaults standardUserDefaults] synchronize];
}


#pragma mark - Networking Indicator

- (void)networkingIndicatorDidStartWithType:(enum NetworkingIndicatorType)type {
    if (!self.screenshotsViewController.navigationItem.leftBarButtonItem) {
        UIActivityIndicatorView *activityView = [[UIActivityIndicatorView alloc] initWithActivityIndicatorStyle:UIActivityIndicatorViewStyleWhite];
        activityView.color = [UIColor crazeRed];
        [activityView startAnimating];
        
        self.screenshotsViewController.navigationItem.leftBarButtonItem = [[UIBarButtonItem alloc] initWithCustomView:activityView];
    }
    
    self.screenshotsViewController.navigationItem.leftBarButtonItem.tag += 1;
}

- (void)networkingIndicatorDidCompleteWithType:(enum NetworkingIndicatorType)type {
    self.screenshotsViewController.navigationItem.leftBarButtonItem.tag -= 1;
    
    if (self.screenshotsViewController.navigationItem.leftBarButtonItem.tag == 0) {
        self.screenshotsViewController.navigationItem.leftBarButtonItem = nil;
    }
}


#pragma mark - Clip View

- (void)presentPickerClipView {
    if (!self.clipView) {
        UIView *rightBarButtonView = self.screenshotsViewController.navigationItem.rightBarButtonItem.targetView;
        CGRect rect = [rightBarButtonView.superview convertRect:rightBarButtonView.frame toView:self.view];
        CGFloat radius = MIN(rect.size.height / 2.f, rect.size.width / 2.f);
        UIBezierPath *croppedPath = [UIBezierPath bezierPathWithRoundedRect:rect cornerRadius:radius];
        
        UIView *tabBarView = self.tabBarController.view;
        
        ClipView *clipView = [[ClipView alloc] init];
        clipView.translatesAutoresizingMaskIntoConstraints = NO;
        clipView.clippings = @[croppedPath];
        clipView.alpha = 0.f;
        [tabBarView addSubview:clipView];
        [clipView.topAnchor constraintEqualToAnchor:tabBarView.topAnchor].active = YES;
        [clipView.leadingAnchor constraintEqualToAnchor:tabBarView.leadingAnchor].active = YES;
        [clipView.bottomAnchor constraintEqualToAnchor:tabBarView.bottomAnchor].active = YES;
        [clipView.trailingAnchor constraintEqualToAnchor:tabBarView.trailingAnchor].active = YES;
        self.clipView = clipView;
        
        [UIView animateWithDuration:Constants.defaultAnimationDuration animations:^{
            self.clipView.alpha = 1.f;
        }];
    }
}

- (void)dismissPickerClipView {
    if (self.clipView) {
        [UIView animateWithDuration:Constants.defaultAnimationDuration animations:^{
            self.clipView.alpha = 0.f;
            
        } completion:^(BOOL finished) {
            [self.clipView removeFromSuperview];
            self.clipView = nil;
        }];
    }
}

#pragma mark - State Restoration

- (void)encodeRestorableStateWithCoder:(NSCoder *)coder {
    ProductsViewController *productsViewController = (ProductsViewController *)self.topViewController;
    
    if ([productsViewController isKindOfClass:[ProductsViewController class]]) {
        NSInteger index = [self.screenshotsViewController indexForScreenshot:productsViewController.screenshot];
        [coder encodeInteger:index forKey:@"screenshotIndex"];
    }
    
    [super encodeRestorableStateWithCoder:coder];
}

- (void)decodeRestorableStateWithCoder:(NSCoder *)coder {
    self.restoredScreenshotNumber = @([coder decodeIntegerForKey:@"screenshotIndex"]);
    
    [super decodeRestorableStateWithCoder:coder];
}

- (void)applicationFinishedRestoringState {
    [super applicationFinishedRestoringState];
    
//    if (self.restoredScreenshotNumber) {
//        ProductsViewController *productsViewController = (ProductsViewController *)self.topViewController;
//
//        if ([productsViewController isKindOfClass:[ProductsViewController class]]) {
//            productsViewController.screenshot = [self.screenshotsViewController screenshotAtIndex:self.restoredScreenshotNumber.integerValue];
//            self.restoredScreenshotNumber = nil;
//        }
//    }
}

@end

File: screenshot/View/Screenshot/ScreenshotsNavigationController.swift
//
//  ScreenshotsNavigationController.swift
//  screenshot
//
//  Created by Jonathan Rose on 2/21/18.
//  Copyright (c) 2018 crazeapp. All rights reserved.
//

import UIKit
import CoreData

protocol ScreenshotsNavigationControllerDelegate: NSObjectProtocol {
    func screenshotsNavigationControllerDidGrantPushPermissions(_ navigationController: ScreenshotsNavigationController)
}

class ScreenshotsNavigationController: UINavigationController {
    weak var screenshotsNavigationControllerDelegate:ScreenshotsNavigationControllerDelegate?
    var screenshotsViewController:ScreenshotsViewController = ScreenshotsViewController()
    fileprivate var restoredProductsViewController: ProductsViewController?
    
    
    init() {
        super.init(nibName: nil, bundle: nil)
        screenshotsViewController.navigationItem.rightBarButtonItem = screenshotsViewController.editButtonItem
        screenshotsViewController.delegate = self
        
        self.viewControllers = [self.screenshotsViewController]
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.view.backgroundColor = .background
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        UniversalSearchController.shared.updateInboxBadgeCount()
    }
}

extension ScreenshotsNavigationController: ViewControllerLifeCycle {
    func viewController(_ viewController: UIViewController, didDisappear animated: Bool) {
        if let productsViewController = viewController as? ProductsViewController, let _ = self.topViewController as? ScreenshotsViewController {
            if needsToPresentPushAlert {
                presentPushAlert()
            }
            
            let screenshot = productsViewController.screenshot
            AssetSyncModel.sharedInstance.clearSubShoppables(screenshot: screenshot)
        }
    }
}

typealias ScreenshotsNavigationControllerPicker = ScreenshotsNavigationController
extension ScreenshotsNavigationControllerPicker {
    func createScreenshotPickerNavigationController()->ScreenshotPickerNavigationController{
        let navigationController = ScreenshotPickerNavigationController.init(nibName: nil, bundle: nil)
        navigationController.cancelButton.target = self
        navigationController.cancelButton.action = #selector(pickerViewControllerDidCancel)
        navigationController.doneButton.target = self
        navigationController.doneButton.action = #selector(pickerViewControllerDidFinish)
        return navigationController
    }
    
    func presentPickerViewController(openScreenshots:Bool) {
        let picker = self.createScreenshotPickerNavigationController()
        picker.screenshotPickerViewController.openScreenshots = openScreenshots
        self.present(picker, animated: true, completion: nil)
        
        Analytics.trackOpenedPicker()
    }
    
    @objc func pickerViewControllerDidCancel() {
        self.dismiss(animated: true)
    }
    
    @objc func pickerViewControllerDidFinish(){
        self.dismiss(animated: true, completion: nil)
    }
}

extension ScreenshotsNavigationController :ScreenshotsViewControllerDelegate{
    func screenshotsViewController(_ viewController:ScreenshotsViewController, didSelectItemAt:IndexPath) {
        if viewController == self.topViewController {
            if let screenshot = viewController.screenshot(at: didSelectItemAt.item) {
                presentScreenshot(screenshot)
            }
        }
    }
    
    func presentScreenshot(_ screenshot: Screenshot, animated:Bool = true) {
        
        let productsViewController = createProductsViewController(screenshot: screenshot)
        self.pushViewController(productsViewController, animated: animated)
        
        if (screenshot.isNew) {
            if screenshot.source == .discover {
                AccumulatorModel.screenshotUninformed.decrementUninformedCount(by:1)
            }
            screenshot.setViewed()
        }
    }
    
    func screenshotsViewControllerWantsToPresentPicker(_  viewController:ScreenshotsViewController, openScreenshots:Bool){
        self.presentPickerViewController(openScreenshots:openScreenshots)
    }
}



typealias ScreenshotsNavigationControllerProducts = ScreenshotsNavigationController
extension ScreenshotsNavigationControllerProducts {
    func createProductsViewController(screenshot: Screenshot) -> ProductsViewController {
        let productsViewController = ProductsViewController(screenshot: screenshot)
        productsViewController.lifeCycleDelegate = self
        return productsViewController
    }
    
    func createRestoredProductsViewController() -> ProductsViewController? {
        // This solution does not work. The only solution is to design the ProductsViewController so it can be created without a screenshot. In the meantime, return nil and restoration wont work for this view controller.
        return nil
//        let tempScreenshot = Screenshot()
//        let productsViewController = createProductsViewController(screenshot: tempScreenshot)
//        restoredProductsViewController = productsViewController
//        return productsViewController
    }
}

typealias ScreenshotsNavigationControllerPushPermission = ScreenshotsNavigationController
extension ScreenshotsNavigationControllerPushPermission {
    fileprivate var needsToPresentPushAlert: Bool {
        return !UserDefaults.standard.bool(forKey: UserDefaultsKeys.onboardingPresentedPushAlert) && !PermissionsManager.shared.hasPermission(for: .push)
    }
    
    fileprivate func presentPushAlert() {
        let alertController = UIAlertController.init(title: "screenshot.permission.push.title".localized, message: "screenshot.permission.push.message".localized, preferredStyle: .alert)
        alertController.addAction(UIAlertAction.init(title: "generic.ok".localized, style: .default, handler: { action in
            PermissionsManager.shared.requestPermission(for: .push, response: { granted in
                if granted {
                    self.screenshotsNavigationControllerDelegate?.screenshotsNavigationControllerDidGrantPushPermissions(self)
                    Analytics.trackAcceptedPushPermissions()
                }
                else {
                    Analytics.trackDeniedPushPermissions()
                }
            })
        }))
        present(alertController, animated: true, completion: nil)
        
        UserDefaults.standard.set(true, forKey: UserDefaultsKeys.onboardingPresentedPushAlert)
        UserDefaults.standard.synchronize()
    }
}
